src/libsyntax/parse/parser.rs

```diff
@@ -5127,7 +5127,11 @@ impl<'a> Parser<'a> {
         let mut fields = Vec::new();
         if self.eat(&token::OpenDelim(token::Brace)) {
             while self.token != token::CloseDelim(token::Brace) {
-                fields.push(self.parse_struct_decl_field()?);
+                fields.push(self.parse_struct_decl_field().map_err(|e| {
+                    self.recover_stmt();
+                    self.eat(&token::CloseDelim(token::Brace));
```

Comment:
Perhaps I should make this into a `match` to make it clearer. If there's an error then we'll hit an early return out of the function (`?` below).